### DISCUSSION TITLE
Notification rejected by the APNs gateway: Optional[InvalidProviderToken]

### DISCUSSION THREAD

**Discussion by suganyagladys** [relayrides/pushy#915] [Q&A]
Hi, 

I am able to send Push Notification with the same KeyId,TeamId and .p8 file from PN tester. Running the java code for PN i am getting the response as  Notification rejected by the APNs gateway: Optional[InvalidProviderToken]

**Comment by jchambers** [comment]
Respectfully, issues like these get opened from time to time and they invariably get closed with something like "I accidentally switched the key ID and team ID." I'm happy to help, but strongly, strongly suspect this isn't a bug that we can fix with a change to Pushy itself. With that in mind, I'm going to convert this to a discussion.

**Reply by jchambers** [reply]
…with that said:

1. Have you tried swapping the key ID and team ID?
2. Have you made any modifications to the .p8 file (i.e. converted it to/from another format or manually edited the file in any way)?

**Reply by suganyagladys** [reply]
Hi, 

1) Have you tried swapping the key ID and team ID? I tried swapping and used the correct  one mentioned as per doc
 .setSigningKey(ApnsSigningKey.loadFromPkcs8File(new File("/path/to/key.p8"),
                "TEAMID1234", "KEYID67890"))
2) Have you made any modifications to the .p8 file (i.e. converted it to/from another format or manually edited the file in any way)? No didn't do any changes.

**Reply by suganyagladys** [reply]
.. Added error response. Please help me to find the root cause

```
SimplePushNotificationResponse{pushNotification=SimpleApnsPushNotification{token='6ff6319f5a6d5eff6b2db0df9d88075799c88a90ec2b60caef12c309e4c86b5a', payload='com.turo.pushy.apns.util.ApnsPayloadBuilder@45fb52b3', invalidationTime=Tue Sep 07 17:31:49 IST 2021, priority=IMMEDIATE, pushType=null, topic='com.sample.PushAPN', collapseId='null', apnsId=null}, success=false, apnsId=46b29324-7032-88ed-bd31-d6e6ee18a7c0, rejectionReason='InvalidProviderToken', tokenExpirationTimestamp=null}
```

**Comment by jchambers** [comment]
One thing I'm noticing is that you appear to be passing in the payload _builder_ instead of the payload itself (perhaps you're calling `toString()` instead of [`build()`](https://pushy-apns.org/apidocs/latest/com/eatthepath/pushy/apns/util/ApnsPayloadBuilder.html#build())?). I don't _think_ that's likely to be the problem (and I don't think the APNs servers actually look at the payload at all), but there's some possibility that there's some other error getting mis-classified upstream. Let's get that sorted and see if the problem persists.

Beyond that, I'm assuming that you've changed the topic from your actual topic to `com.sample.PushAPN` for the purposes of sharing an example. If that's _not_ the case, though, I think it'll be important to try again with the correct topic for your target app. Otherwise, the servers may be (correctly!) recognizing that your provider token is not associated with `com.sample.PushAPN` as a topic.

**Comment by suganyagladys** [comment]
Hi, Instead of .toString(). I changed to below format

final ApnsPayloadBuilder payloadBuilder = new SimpleApnsPayloadBuilder();
				 payloadBuilder.setAlertBody("Example!");
				 System.out.println("Build payloadBuilder"); 
				    final String payload = payloadBuilder.build();
				    final String token = TokenUtil.sanitizeTokenString("6ff6319f5a6d5eff6b2db0df9d88075799c88a90ec2b60caef12c309e4c86b5a");
 

FAILURESimplePushNotificationResponse{pushNotification=SimpleApnsPushNotification{token='6ff6319f5a6d5eff6b2db0df9d88075799c88a90ec2b60caef12c309e4c86b5a', payload='{"aps":{"alert":{"body":"Example!"}}}', invalidationTime=2021-09-08T07:44:20.888Z, priority=IMMEDIATE, pushType=null, topic='com.sample.PushAPN', collapseId='null', apnsId=null}, success=false, apnsId=dbc5e598-6138-562c-5707-89d3f6016afe, rejectionReason='InvalidProviderToken', tokenExpirationTimestamp=null}

I am able to send PN using cert file. the same way code i am using for Payload builder there it is working. 
Yes,I changed the topic name here.
